File: src/services/audit/detectors/ad/groups/size/oversized-critical.ts
/**
 * OVERSIZED_GROUP_CRITICAL - Groups with 500+ members
 */

import { ADGroup } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { toAffectedADGroupEntities } from '../../../../../../utils/entity-converter';

/**
 * Check for oversized groups (500+ members)
 * PingCastle threshold: >500 members = critical
 */
export function detectOversizedGroupCritical(groups: ADGroup[], includeDetails: boolean): Finding {
  const affected = groups.filter((g) => {
    if (!g.member) return false;
    return g.member.length > 500;
  });

  return {
    type: 'OVERSIZED_GROUP_CRITICAL',
    severity: 'high',
    category: 'groups',
    title: 'Oversized Group (Critical)',
    description: 'Groups with 500+ members. Management/audit difficulty, excessive privileges, performance issues.',
    count: affected.length,
    affectedEntities: includeDetails ? toAffectedADGroupEntities(affected) : undefined,
  };
}

File: internal/audit/detectors/ad/groups/size/oversized-critical.go
package size

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// OversizedCriticalDetector checks for groups with 500+ members
type OversizedCriticalDetector struct {
	audit.BaseDetector
}

// NewOversizedCriticalDetector creates a new detector
func NewOversizedCriticalDetector() *OversizedCriticalDetector {
	return &OversizedCriticalDetector{
		BaseDetector: audit.NewBaseDetector("OVERSIZED_GROUP_CRITICAL", audit.CategoryGroups),
	}
}

// Detect executes the detection
func (d *OversizedCriticalDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	var affected []string

	for _, group := range data.Groups {
		if len(group.Member) > 500 {
			name := group.SAMAccountName
			if name == "" {
				name = group.DistinguishedName
			}
			affected = append(affected, name)
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityHigh,
		Category:    string(d.Category()),
		Title:       "Oversized Group (Critical)",
		Description: "Groups with 500+ members. Management/audit difficulty, excessive privileges, performance issues.",
		Count:       len(affected),
	}

	if len(affected) > 0 {
		finding.AffectedEntities = toAffectedGroupEntities(affected)
	}

	return []types.Finding{finding}
}

func toAffectedGroupEntities(names []string) []types.AffectedEntity {
	entities := make([]types.AffectedEntity, len(names))
	for i, name := range names {
		entities[i] = types.AffectedEntity{
			Type: "group",
			Name: name,
		}
	}
	return entities
}

func init() {
	audit.MustRegister(NewOversizedCriticalDetector())
}
